### PR TITLE
add more filter options

### PR DESCRIPTION
**Pull request by rbellens**
* filter devices on name and name prefix
* add optional services property
* allow to define filters and options in request method argument

### CODE REVIEW

**Comment by beaufortfrancois** [comment]
I was waiting for this to land in Chrome OS Dev Channel but I'm glad you did it first. 
Will review this shortly if that's ok with you!

Thank you!

**Comment by rbellens** [comment]
Ok, thanks.

I am using the chrome app polyfill with mobile chrome apps (see also WebBluetoothCG/chrome-app-polyfill#6).

**Comment by beaufortfrancois** [comment]
May I ask why mobile chrome apps instead of a regular mobile wen app?

**Comment by rbellens** [comment]
because we need bluetooth and possibly other native features in the future, which we can access through cordova plugins

**Comment by beaufortfrancois** [comment]
Oh it does make sense then ;)

File: platinum-bluetooth-device.html [hunk 1]
```diff
@@ -83,7 +83,33 @@
...
+          type: String,
+          observer: '_filterChanged'
+        },
+
+        /**
+         * Optional Bluetooth GATT services to get access to. You may provide
+         * either the full Bluetooth UUID as a string or a short 16- or 32-bit
+         * form as integers like 0x180d.
+         */
+        optionalServices: {
```
Comment (beaufortfrancois) [inline platinum-bluetooth-device.html]: Could you remove optionalServices as it's not implemented yet?
I just want to lower down developers expectation there ;)

File: platinum-bluetooth-device.html [hunk 2]
```diff
@@ -139,7 +165,7 @@
        *
        * @return {Promise<BluetoothDevice>}
        */
-      request: function() {
+      request: function(options) {
```
Comment (beaufortfrancois) [inline platinum-bluetooth-device.html]: Can you add a comment above to specify what is the `options` object?

File: platinum-bluetooth-device.html [hunk 3]
```diff
@@ -148,11 +174,24 @@
           // Resolve promise if user already picked one device.
           return Promise.resolve(this._device);
         }
-        if (!this.servicesFilter || this.servicesFilter.length == 0) {
-          return Promise.reject(new Error('Please set Bluetooth services filter.'));
+        options = options||{};
```
Comment (beaufortfrancois) [inline platinum-bluetooth-device.html]: `s/options||{}/options\ ||\ {}`

File: platinum-bluetooth-device.html [hunk 4]
```diff
@@ -148,11 +174,24 @@
...
+        if ((!this.servicesFilter || this.servicesFilter.length == 0)
+                &&!this.nameFilter&&!this.namePrefixFilter) {
```
Comment (beaufortfrancois) [inline platinum-bluetooth-device.html]: Same whitespace issue there as above.

File: platinum-bluetooth-device.html [hunk 5]
```diff
@@ -148,11 +174,24 @@
...
+          if (!(options.filters&&options.filters.length)) {
```
Comment (beaufortfrancois) [inline platinum-bluetooth-device.html]: whitespace issue there as well.

File: platinum-bluetooth-device.html [hunk 6]
```diff
@@ -148,11 +174,24 @@
...
+            return Promise.reject(new Error('Please set Bluetooth services, name and/or name prefix filter.'));
+          }
+        } else {
+          options.filters = options.filters||[];
```
Comment (beaufortfrancois) [inline platinum-bluetooth-device.html]: whitespace issue there as well.

File: platinum-bluetooth-device.html [hunk 7]
```diff
@@ -148,11 +174,24 @@
...
+          options.filters.push({
+            services: this.servicesFilter,
+            name: this.nameFilter,
+            namePrefix: this.namePrefixFilter
+          })
         }
+        options.optionalServices = this.optionalServices||options.optionalServices;
```
Comment (beaufortfrancois) [inline platinum-bluetooth-device.html]: Can we remove it since we don't want to use it for now?

File: platinum-bluetooth-device.html [hunk 8]
```diff
@@ -148,11 +174,24 @@
...
-          return Promise.reject(new Error('Please set Bluetooth services filter.'));
+        options = options||{};
+        if ((!this.servicesFilter || this.servicesFilter.length == 0)
+                &&!this.nameFilter&&!this.namePrefixFilter) {
+          if (!(options.filters&&options.filters.length)) {
+            return Promise.reject(new Error('Please set Bluetooth services, name and/or name prefix filter.'));
+          }
+        } else {
+          options.filters = options.filters||[];
+          options.filters.push({
```
Comment (beaufortfrancois) [inline platinum-bluetooth-device.html]: We may not want to populate `name` for instance if `nameFilter` is not defined.

**Comment by beaufortfrancois** [comment]
@rbellens I've added some minor comments to your PR. Thank you!

**Comment by rbellens** [comment]
@beaufortfrancois I made the changes according to your comments

**Comment by beaufortfrancois** [comment]
Thank you @rbellens and sorry for the late reply ;)

After giving it more thoughts, I think we may want to remove the `options` parameters and simply use the new `name-filter` and `name-prefix-filter` attributes you've already added when requesting a device.

It would make the code simpler and avoid two ways of requesting a device.
What do you think?

**Comment by rbellens** [comment]
The `options` parameter gives more flexibility. For example to add an `optionalServices` parameter or possible other `RequestDeviceOptions` that might be added to the standard. 
For my project, I wanted to be able to connect to a device without the user prompt, for which I added a `silent` parameter (see appsup/chrome-app-polyfill@4e922699b28ba8090e716e9333d7f30be1de5366).
I would understand however, if you prefer sticking to the standard and keeping it simple.

**Comment by beaufortfrancois** [comment]
I've eventually decided to keep it simple for the next version of `<platinum-bluetooth>` elements: https://github.com/PolymerElements/platinum-bluetooth/commit/6be9e7aab3bf55b4199e0772c84ea5b5a4ca096b

**Pull request closed without merging**